system/Files/FileCollection.php

```diff
@@ -0,0 +1,396 @@
+ */
+class FileCollection implements Countable, IteratorAggregate
+{
+	/**
+	 * The current list of file paths.
+	 *
+	 * @var string[]
+	 */
+	protected $files = [];
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Resolves a full path and verifies it is an actual directory.
+	 *
+	 * @param string $directory
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	final protected static function resolveDirectory(string $directory): string
+	{
+		if (! is_dir($directory = set_realpath($directory)))
+		{
+			$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1];
+			throw FileException::forExpectedDirectory($caller['function']);
+		}
+
+		return $directory;
+	}
+
+	/**
+	 * Resolves a full path and verifies it is an actual file.
+	 *
+	 * @param string $file
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	final protected static function resolveFile(string $file): string
+	{
+		if (! is_file($file = set_realpath($file)))
+		{
+			$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1];
+			throw FileException::forExpectedFile($caller['function']);
+		}
+
+		return $file;
+	}
+
+	/**
+	 * Removes files that are not part of the given directory (recursive).
+	 *
+	 * @param string[] $files
+	 * @param string   $directory
+	 *
+	 * @return string[]
+	 */
+	final protected static function filterFiles(array $files, string $directory): array
+	{
+		$directory = self::resolveDirectory($directory);
+
+		return array_filter($files, function ($value) use ($directory) {
+			return strpos($value, $directory) === 0;
+		});
+	}
+
+	/**
+	 * Returns any files whose `basename` matches the given pattern.
+	 *
+	 * @param string[] $files
+	 * @param string   $pattern Regex or pseudo-regex string
+	 *
+	 * @return string[]
+	 */
+	final protected static function matchFiles(array $files, string $pattern): array
+	{
+		// Convert pseudo-regex into their true form
+		if (@preg_match($pattern, null) === false) // @phpstan-ignore-line
+		{
+			$pattern = str_replace(
+				['#', '.', '*', '?'],
+				['\#', '\.', '.*', '.'],
+				$pattern
+			);
+			$pattern = "#{$pattern}#";
+		}
+
+		return array_filter($files, function ($value) use ($pattern) {
+			return (bool) preg_match($pattern, basename($value));
+		});
+	}
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Loads the Filesystem helper and adds any initial files.
+	 *
+	 * @param string[] $files
+	 */
+	public function __construct(array $files = [])
+	{
+		helper(['filesystem']);
+
+		$this->add($files)->define();
+	}
+
+	/**
+	 * Applies any initial inputs after the constructor.
+	 * This method is a stub to be implemented by child classes.
+	 *
+	 * @return void
+	 */
+	protected function define(): void
+	{
+	}
+
+	/**
+	 * Optimizes and returns the current file list.
+	 *
+	 * @return string[]
+	 */
+	public function get(): array
+	{
+		$this->files = array_unique($this->files, SORT_STRING);
```

Comment:
```suggestion
		$this->files = array_unique($this->files);
```
`SORT_STRING` is the default flag.